os/arch/arm/src/amebad/amebad_timer_lowerhalf.c

```diff
@@ -0,0 +1,485 @@
+#include "timer_api.h"
+
+#ifdef CONFIG_TIMER
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+struct amebad_gpt_lowerhalf_s {
+	const struct timer_ops_s *ops;	/* Lowerhalf operations */
+	gtimer_t obj;
+	bool started;				/* True: Timer is started */
+	tccb_t callback;
+	void *arg;					/* Argument passed to upper half callback */
+	uint32_t gpt_timeout;
+	bool freerunmode;			/* True: Free Run Mode. False: Alarm Mode */
+	uint32_t freeruncount;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+static int amebad_gpt_start(struct timer_lowerhalf_s *lower);
+static int amebad_gpt_stop(struct timer_lowerhalf_s *lower);
+static int amebad_gpt_getstatus(struct timer_lowerhalf_s *lower, struct timer_status_s *status);
+static int amebad_gpt_settimeout(struct timer_lowerhalf_s *lower, uint32_t timeout);
+static void amebad_gpt_setcallback(struct timer_lowerhalf_s *lower, CODE tccb_t callback, void *arg);
+static int amebad_gpt_ioctl(struct timer_lowerhalf_s *lower, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+static const struct timer_ops_s g_timer_ops = {
+	.start = amebad_gpt_start,
+	.stop = amebad_gpt_stop,
+	.getstatus = amebad_gpt_getstatus,
+	.settimeout = amebad_gpt_settimeout,
+	.setcallback = amebad_gpt_setcallback,
+	.ioctl = amebad_gpt_ioctl,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt0_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt1_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt2_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt3_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: amebad_gpt_handler
+ *
+ * Description:
+ *   timer interrupt handler
+ *
+ * Input Parameters:
+ *
+ * Returned Values:
+ *
+ ****************************************************************************/
+static void amebad_gpt_handler(uint32_t data)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)data;
+	DEBUGASSERT(priv);
+	uint32_t next_interval_us = 0;
+
+	if (priv->freerunmode) {	/* Free Run Mode */
+		priv->freeruncount++;
```

Comment:
Could you explain the meaning of this logic? Is there any case getting interrupt with free-run mode?

Reply:
As our timers do not have the free run mode setting, the free running feature has to be implemented in this logic. So that, maximum timeout is firstly set to that free run timer, when it timeout, the interrupt is triggered just to record the times of timeout. Then, the freeruncount is used to calculate the total time elapse.

Reply:
Let me summarize.
1. RTL8721 hw timer does not support free-run mode.
2. Even though 1st, you will support that in sw driver.
   When user sets it as free-run timer, amebad timer driver runs alam mode with maximum timeout value.
   When user calls get status function to get the running time, driver gets the value from alarm mode register and calculates how many times it is expired to get total time.

Is this right?

If it is, let's rename `freeruncount` like `num_of_xxxms_expired`. `xxx` could be calculated using your maximum value.

Reply:
Yes, your understanding is right! I will change the name and add description for it. The max value can be set as 0xFFFFFFFF, so that the name as you suggest will be 'num_of_4294967ms_elapse'. Thanks!